src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,523 @@
+        /// <returns>Returns the full path to the process executable</returns>
+        internal static string proc_pidpath(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid", SR.NegativePidNotSupported);
+            }
+
+            // Allocate our unmanaged buffer for the path
+            string path = string.Empty;
+            uint size = (uint)Marshal.SizeOf<char>() * PROC_PIDPATHINFO_MAXSIZE + 1;
+            IntPtr buffer = Marshal.AllocCoTaskMem((int)size);
+
+            // Get the PID's executable's path
+            int result = proc_pidpath(pid, buffer, size);
+            if (result < 0)
+            {
+                // Check for less than 0 since the kernel process won't have a path
+                throw new COMException(string.Format(SR.CantFindProcessExecutablePath, pid), Marshal.GetLastWin32Error());
+            }
+
+            // Copy unmanaged data into a managed string and free the unmanaged buffer
+            path = Marshal.PtrToStringUni(buffer, result);
+            Marshal.FreeCoTaskMem(buffer);
+
+            return path;
+        }
+
+        /// <summary>
+        /// Gets the rusage information for the process identified by the PID
+        /// </summary>
+        /// <param name="pid">The process to retrieve the rusage for</param>
+        /// <param name="flavor">Should be RUSAGE_SELF to specify getting the info for the specified process</param>
+        /// <param name="rusage_info_t">A buffer to be filled with rusage_info data</param>
+        /// <returns>Returns 0 on success; on fail, -1 and errno is set with the error code</returns>
+        /// <remarks>
+        /// We need to use IntPtr here for the buffer since the function signature uses 
+        /// void* and not a strong type even though it returns a rusage_info struct
+        /// </remarks>
+        [DllImport(Interop.Libraries.libproc, SetLastError = true)]
+        private static extern int proc_pid_rusage(
+            int pid,
+            int flavor,
+            IntPtr rusage_info_t);
+
+        /// <summary>
+        /// Gets the rusage information for the process identified by the PID
+        /// </summary>
+        /// <param name="pid">The process to retrieve the rusage for</param>
+        /// <returns>On success, returns a struct containing info about the process; on
+        /// failure or when the caller doesn't have permissions to the process, throws a COMException
+        /// </returns>
+        internal static rusage_info_v3 proc_pid_rusage(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid", SR.NegativePidNotSupported);
+            }
+
+            // Allocate our managed buffer for the data
+            rusage_info_v3 info;
+            int size = Marshal.SizeOf<rusage_info_v3>();
```

Comment:
The rusage_info_v3 struct is defined using marshaling attributes, in particular to handle an inline array. One small advantage of a pattern that instead used a fixed-size buffer in the struct is that you could avoid these alloc/free calls, and instead just declare the instance on the stack and pass a pointer to it.  That does have downstream consequences, though, as the inline array will then be a pointer rather than a managed array, and the struct itself will need to be unsafe.  Just something to think about; may not make sense here.